central/cloudsources/service/service_impl.go

```diff
@@ -0,0 +1,271 @@
+}
+
+// AuthFuncOverride specifies the auth criteria for this API.
+func (s *serviceImpl) AuthFuncOverride(ctx context.Context, fullMethodName string) (context.Context, error) {
+	return ctx, authorizer.Authorized(ctx, fullMethodName)
+}
+
+// CountCloudSources returns the number of cloud sources matching the request query.
+func (s *serviceImpl) CountCloudSources(ctx context.Context, request *v1.CountCloudSourcesRequest,
+) (*v1.CountCloudSourcesResponse, error) {
+	query := getQueryBuilderFromFilter(request.GetFilter()).ProtoQuery()
+	count, err := s.ds.CountCloudSources(ctx, query)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to count cloud sources")
+	}
+	return &v1.CountCloudSourcesResponse{Count: int32(count)}, nil
+}
+
+// GetCloudSource returns a specific cloud source based on its ID.
+func (s *serviceImpl) GetCloudSource(ctx context.Context, request *v1.GetCloudSourceRequest,
+) (*v1.GetCloudSourceResponse, error) {
+	resourceID := request.GetId()
+	if resourceID == "" {
+		return nil, errors.Wrap(errox.InvalidArgs, "id must be provided")
+	}
+	cloudSource, err := s.ds.GetCloudSource(ctx, resourceID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get cloud source %q", resourceID)
+	}
+	return &v1.GetCloudSourceResponse{CloudSource: toV1Proto(cloudSource)}, err
+}
+
+// ListCloudSources returns all cloud sources matching the request query.
+func (s *serviceImpl) ListCloudSources(ctx context.Context, request *v1.ListCloudSourcesRequest,
+) (*v1.ListCloudSourcesResponse, error) {
+	query := getQueryBuilderFromFilter(request.GetFilter()).ProtoQuery()
+	paginated.FillPagination(query, request.GetPagination(), maxPaginationLimit)
+	paginated.FillDefaultSortOption(
+		query,
+		&v1.QuerySortOption{
+			Field: search.IntegrationName.String(),
+		},
+	)
+
+	cloudSources, err := s.ds.ListCloudSources(ctx, query)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list cloud sources")
+	}
+	v1CloudSources := make([]*v1.CloudSource, 0, len(cloudSources))
+	for _, cs := range cloudSources {
+		v1CloudSources = append(v1CloudSources, toV1Proto(cs))
+	}
+	return &v1.ListCloudSourcesResponse{CloudSources: v1CloudSources}, nil
+}
+
+// PostCloudSource creates a new cloud source.
+func (s *serviceImpl) PostCloudSource(ctx context.Context, request *v1.PostCloudSourceRequest,
+) (*v1.PostCloudSourceResponse, error) {
+	v1CloudSource := request.GetCloudSource()
+	if v1CloudSource.GetId() != "" {
+		return nil, errors.Wrap(errox.InvalidArgs, "id field must be empty when posting a new cloud source")
+	}
+	if err := s.validateCloudSource(ctx, v1CloudSource, true); err != nil {
+		return nil, errors.Wrap(errox.InvalidArgs, err.Error())
+	}
+	v1CloudSource.Id = uuid.NewV4().String()
+	storageCloudSource := toStorageProto(v1CloudSource)
+	if err := s.ds.UpsertCloudSource(ctx, storageCloudSource); err != nil {
+		_ = s.ds.DeleteCloudSource(ctx, storageCloudSource.GetId())
```

Comment:
I also copied this approach from image integrations. I think you are right that we can rely on the store to implement transaction correctly, I'll remove it.

Reply:
Ah this is here for image integrations because there is an intermediate structure that could be added to before the DB so it should not be required in this case